src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
@AArnott ```StreamJsonRpc``` can dispose of the ```DelimitedMessageHandler``` in multiple cases, including closed stream or bogus messages. If another client is using it to send a message on another thread, this may throw ```ArgumentNullException``` like we see in the Roslyn bug. This is not part of the contract of ```StreamJsonRpc``` Invoke* and Notify* methods. In such conditions, this must be ```OperationCanceledException``` throw from these methods.

Now the fix can be either this PR, or a fix in StreamJsonRpc not to dispose ```SemaphoreSlim```.